probe-rs/src/architecture/riscv/mod.rs

```diff
@@ -27,39 +31,39 @@ pub mod communication_interface;
 pub(crate) mod dtm;
 pub mod sequences;
 
-/// An interface to operate RISC-V cores.
+/// An interface to operate a RISC-V core.
 pub struct Riscv32<'probe> {
-    hart: u32,
     interface: &'probe mut RiscvCommunicationInterface,
     state: &'probe mut RiscVState,
+    sequence: Arc<dyn RiscvDebugSequence>,
 }
 
 impl<'probe> Riscv32<'probe> {
-    /// Create a new RISC-V interface.
+    /// Create a new RISC-V interface for a particular hart.
     pub fn new(
         hart: u32,
         interface: &'probe mut RiscvCommunicationInterface,
         state: &'probe mut RiscVState,
+        sequence: Arc<dyn RiscvDebugSequence>,
     ) -> Result<Self, RiscvError> {
         if !interface.hart_enabled(hart) {
             return Err(RiscvError::HartUnavailable);
         }
 
+        interface.select_hart(hart)?;
```

Comment:
This object is created by Session::attach and I don't think it's possible to access multiple cores at the same time. So selecting the hart once should be sufficient and allows removing some boilerplate.